File: backend/app/schemas/analytics.py
"""Analytics API response schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class AnalyticsOverview(BaseModel):
    project_count: int = 0
    total_nodes: int = 0
    total_edges: int = 0
    unique_models: int = 0
    total_files: int = 0
    active_pipelines: int = 0
    total_runs: int = 0
    runs_today: int = 0
    pages_processed: int = 0
    success_rate: float | None = None
    last_activity_at: datetime | None = None


class ActivityBucket(BaseModel):
    bucket_start: datetime
    runs: int = 0
    pages: int = 0
    errors: int = 0
    active_projects: int = 0


class ActivitySeries(BaseModel):
    bucket: str
    from_at: datetime
    to_at: datetime
    items: list[ActivityBucket]


class ModelUsageItem(BaseModel):
    model_id: str
    display_name: str | None = None
    category: str | None = None
    run_count: int = 0
    avg_latency_ms: float | None = None
    success_rate: float | None = None
    last_used_at: datetime | None = None


class ModelUsageList(BaseModel):
    items: list[ModelUsageItem]


class ProjectBreakdownItem(BaseModel):
    project_id: str
    name: str
    node_count: int = 0
    model_count: int = 0
    file_count: int = 0
    run_count: int = 0
    last_activity_at: datetime | None = None
    updated_at: datetime
    owner_id: str | None = None
    owner_email: str | None = None
    icon: str | None = None
    color: str | None = None
    status: str | None = None


class ProjectBreakdownList(BaseModel):
    items: list[ProjectBreakdownItem]


class NodeBreakdownItem(BaseModel):
    project_id: str
    project_name: str
    node_id: str
    model_id: str
    category: str | None = None
    run_status: str | None = None
    last_run_at: str | None = None
    owner_email: str | None = None
    run_count: int = 0


class NodeBreakdownList(BaseModel):
    items: list[NodeBreakdownItem]


class DocumentBreakdownItem(BaseModel):
    project_id: str
    project_name: str
    asset_id: str
    filename: str
    format: str
    size_bytes: int = Field(ge=0)
    uploaded_at: datetime | None = None
    owner_email: str | None = None


class DocumentBreakdownList(BaseModel):
    items: list[DocumentBreakdownItem]


class AnalyticsKpi(BaseModel):
    label: str
    value: str
    delta: str
    delta_direction: str = "neutral"
    delta_label: str


class AnalyticsSummary(BaseModel):
    range: str
    from_at: datetime
    to_at: datetime
    pages_processed: int = 0
    pipeline_runs: int = 0
    success_rate: float | None = None
    avg_latency_ms_per_page: float | None = None
    kpis: list[AnalyticsKpi]


class RunOutcomeSegment(BaseModel):
    label: str
    count: int = 0
    percentage: float = 0.0
    color_key: str


class RunOutcomes(BaseModel):
    total_runs: int = 0
    segments: list[RunOutcomeSegment]


class TopPipelineItem(BaseModel):
    project_id: str
    name: str
    run_count: int = 0
    share: float = 0.0


class TopPipelineList(BaseModel):
    items: list[TopPipelineItem]


class RecentRunItem(BaseModel):
    id: str
    run_label: str
    project_id: str | None = None
    pipeline_name: str
    duration_ms: float | None = None
    status: str
    created_at: datetime
    owner_id: str | None = None
    owner_email: str | None = None


class RecentRunList(BaseModel):
    items: list[RecentRunItem]
    total: int = 0


class UserLeaderboardItem(BaseModel):
    user_id: str
    email: str
    full_name: str | None = None
    display_name: str | None = None
    role: str
    project_count: int = 0
    run_count: int = 0
    pages_processed: int = 0
    last_login_at: datetime | None = None
    last_run_at: datetime | None = None


class UserLeaderboardList(BaseModel):
    items: list[UserLeaderboardItem]


class UserActivityBucket(BaseModel):
    bucket_start: datetime
    active_users: int = 0


class UserActivitySeries(BaseModel):
    bucket: str
    from_at: datetime
    to_at: datetime
    items: list[UserActivityBucket]


class PipelineLibraryStats(BaseModel):
    total_pipelines: int = 0
    active_pipelines: int = 0
    archived_pipelines: int = 0
    avg_nodes: float = 0.0
    avg_models: float = 0.0
    avg_edges: float = 0.0
    unique_io_types: int = 0


class PipelineBreakdownItem(BaseModel):
    pipeline_id: str
    name: str
    description: str | None = None
    node_count: int = 0
    edge_count: int = 0
    model_count: int = 0
    input_type_label: str | None = None
    output_type_label: str | None = None
    accent_color: str = "#5B2EEF"
    is_archived: bool = False
    created_at: datetime
    updated_at: datetime
    owner_id: str | None = None
    owner_email: str | None = None


class PipelineBreakdownList(BaseModel):
    items: list[PipelineBreakdownItem]


class RunKindSegment(BaseModel):
    label: str
    count: int = 0
    percentage: float = 0.0
    color_key: str


class RunKindBreakdown(BaseModel):
    total_runs: int = 0
    segments: list[RunKindSegment]

File: frontend/src/lib/api/analytics.ts
export type AnalyticsRange = "7d" | "30d" | "90d";

export type AnalyticsOverview = {
  project_count: number;
  total_nodes: number;
  total_edges: number;
  unique_models: number;
  total_files: number;
  active_pipelines: number;
  total_runs: number;
  runs_today: number;
  pages_processed: number;
  success_rate: number | null;
  last_activity_at: string | null;
};

export type ActivityBucket = {
  bucket_start: string;
  runs: number;
  pages: number;
  errors: number;
  active_projects: number;
};

export type ActivitySeries = {
  bucket: string;
  from_at: string;
  to_at: string;
  items: ActivityBucket[];
};

export type AnalyticsKpi = {
  label: string;
  value: string;
  delta: string;
  delta_direction: "up" | "down" | "neutral";
  delta_label: string;
};

export type AnalyticsSummary = {
  range: AnalyticsRange;
  from_at: string;
  to_at: string;
  pages_processed: number;
  pipeline_runs: number;
  success_rate: number | null;
  avg_latency_ms_per_page: number | null;
  kpis: AnalyticsKpi[];
};

export type RunOutcomeSegment = {
  label: string;
  count: number;
  percentage: number;
  color_key: "done" | "failed" | "running";
};

export type RunOutcomes = {
  total_runs: number;
  segments: RunOutcomeSegment[];
};

export type TopPipelineItem = {
  project_id: string;
  name: string;
  run_count: number;
  share: number;
};

export type TopPipelineList = {
  items: TopPipelineItem[];
};

export type RecentRunItem = {
  id: string;
  run_label: string;
  project_id: string | null;
  pipeline_name: string;
  duration_ms: number | null;
  status: string;
  created_at: string;
  owner_id?: string | null;
  owner_email?: string | null;
};

export type RecentRunList = {
  items: RecentRunItem[];
  total: number;
};

export type ModelUsageItem = {
  model_id: string;
  display_name: string | null;
  category: string | null;
  run_count: number;
  avg_latency_ms: number | null;
  success_rate: number | null;
  last_used_at: string | null;
};

export type ModelUsageList = {
  items: ModelUsageItem[];
};

export type ProjectBreakdownItem = {
  project_id: string;
  name: string;
  node_count: number;
  model_count: number;
  file_count: number;
  run_count: number;
  last_activity_at: string | null;
  updated_at: string;
  owner_id?: string | null;
  owner_email?: string | null;
  icon?: string | null;
  color?: string | null;
  status?: string | null;
};

export type ProjectBreakdownList = {
  items: ProjectBreakdownItem[];
};

export type NodeBreakdownItem = {
  project_id: string;
  project_name: string;
  node_id: string;
  model_id: string;
  category: string | null;
  run_status: string | null;
  last_run_at: string | null;
  owner_email?: string | null;
  run_count?: number;
};

export type NodeBreakdownList = {
  items: NodeBreakdownItem[];
};

export type DocumentBreakdownItem = {
  project_id: string;
  project_name: string;
  asset_id: string;
  filename: string;
  format: string;
  size_bytes: number;
  uploaded_at: string | null;
  owner_email?: string | null;
};

export type DocumentBreakdownList = {
  items: DocumentBreakdownItem[];
};

export type UserActivityBucket = {
  bucket_start: string;
  active_users: number;
};

export type UserActivitySeries = {
  bucket: string;
  from_at: string;
  to_at: string;
  items: UserActivityBucket[];
};

export type PipelineLibraryStats = {
  total_pipelines: number;
  active_pipelines: number;
  archived_pipelines: number;
  avg_nodes: number;
  avg_models: number;
  avg_edges: number;
  unique_io_types: number;
};

export type PipelineBreakdownItem = {
  pipeline_id: string;
  name: string;
  description: string | null;
  node_count: number;
  edge_count: number;
  model_count: number;
  input_type_label: string | null;
  output_type_label: string | null;
  accent_color: string;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
  owner_id?: string | null;
  owner_email?: string | null;
};

export type PipelineBreakdownList = {
  items: PipelineBreakdownItem[];
};

export type RunKindSegment = {
  label: string;
  count: number;
  percentage: number;
  color_key: "test_run" | "pipeline_run" | "inference_run";
};

export type RunKindBreakdown = {
  total_runs: number;
  segments: RunKindSegment[];
};

export const ANALYTICS_RANGES: AnalyticsRange[] = ["7d", "30d", "90d"];

export const RANGE_LABELS: Record<AnalyticsRange, string> = {
  "7d": "7D",
  "30d": "30D",
  "90d": "90D",
};

export const RANGE_CAPTIONS: Record<AnalyticsRange, string> = {
  "7d": "Last 7 days",
  "30d": "Last 30 days",
  "90d": "Last 90 days",
};

export function formatSuccessRate(value: number | null): string {
  if (value === null) return "—";
  return `${Math.round(value * 100)}%`;
}

export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function formatDurationMs(value: number | null | undefined): string {
  if (value == null) return "—";
  return `${(value / 1000).toFixed(1)}s`;
}

export function buildAnalyticsQuery(
  basePath: string,
  params: Record<string, string | undefined>,
): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value) search.set(key, value);
  }
  const query = search.toString();
  return query ? `${basePath}?${query}` : basePath;
}

export function buildAnalyticsDashboardQuery(
  basePath: string,
  options: {
    range?: AnalyticsRange;
    project_id?: string;
    limit?: number;
  },
): string {
  return buildAnalyticsQuery(basePath, {
    range: options.range ?? "30d",
    project_id: options.project_id,
    limit: options.limit?.toString(),
  });
}

export function getAnalyticsExportUrl(
  range: AnalyticsRange,
  projectId?: string | null,
): string {
  const params = new URLSearchParams({ range });
  if (projectId) params.set("project_id", projectId);
  return `/api/analytics/export?${params.toString()}`;
}

export function resolveActivityRangeDates(range: AnalyticsRange): {
  from: string;
  to: string;
} {
  const to = new Date();
  const from = new Date(to);
  const days = range === "7d" ? 7 : range === "90d" ? 90 : 30;
  from.setDate(from.getDate() - days);
  return { from: from.toISOString(), to: to.toISOString() };
}
